tools/pubsys/src/aws/client.rs

```diff
@@ -1,159 +1,150 @@
-/// opted-in to that region, and you need to get session credentials from an STS endpoint in a
-/// region to which you have access in the base account.
-fn build_provider<P>(
-    sts_region: &Region,
+async fn build_chained_config(
+    region: &str,
     assume_roles: impl Iterator<Item = String>,
-    base_provider: P,
-) -> Result<CredentialsProvider>
-where
-    P: ProvideAwsCredentials + Send + Sync + 'static,
-{
-    let mut provider = CredentialsProvider(Box::new(base_provider));
+    base_config: SdkConfig,
+) -> Result<ConfigLoader> {
+    let mut provider = aws_config::from_env();
+    let region = region.to_owned();
+
+    // Chains credentials providers to assume the given roles in order.
+    // The region given should be the one in which you want to talk to STS to get temporary
+    // credentials, not the region in which you want to talk to a service endpoint like EC2.  This is
+    // needed because you may be assuming a role in an opt-in region from an account that has not
+    // opted-in to that region, and you need to get session credentials from an STS endpoint in a
+    // region to which you have access in the base account
+    let mut sts_config = base_config;
     for assume_role in assume_roles {
-        let sts = StsClient::new_with(
-            HttpClient::new().context(error::HttpClientSnafu)?,
-            provider,
-            sts_region.clone(),
+        let sts_client = StsClient::new(&sts_config);
+        let sts_response = sts_client
+            .assume_role()
+            .role_arn(assume_role.clone())
+            .role_session_name(PROVIDER_NAME)
+            .send()
+            .await
+            .context(error::AssumeRoleSnafu {
+                role_arn: &assume_role,
+            })?;
+        // Parse the StsCredentials from the response.
+        let sts_credentials: &StsCredentials =
+            sts_response
+                .credentials()
+                .context(error::CredentialsMissingSnafu {
+                    credentials_key: "Credentials",
+                    role_arn: &assume_role,
+                })?;
+        // Convert StsCredentials to SdkCredentials.
+        let credentials = SdkCredentials::new(
+            sts_credentials
+                .access_key_id()
+                .context(error::CredentialsMissingSnafu {
+                    credentials_key: "access_key_id",
+                    role_arn: &assume_role,
+                })?,
+            sts_credentials
+                .secret_access_key()
+                .context(error::CredentialsMissingSnafu {
+                    credentials_key: "secret_access_key",
+                    role_arn: &assume_role,
+                })?,
+            sts_credentials.session_token().map(|s| s.to_string()),
+            Some(
+                SystemTime::try_from(sts_credentials.expiration.unwrap())
+                    .context(error::DateTimeToSystemTimeSnafu)?,
+            ),
```

Comment:
Is that a safe `unwrap`? What if it's `None`? Why can't we use the `sts_credentials.expiration` as is?